flowmachine/flowmachine/features/subscriber/distance_counterparts.py

```diff
@@ -0,0 +1,151 @@
+    """
+    This class returns metrics related with the distance between event
+    initiator and her/his counterparts.
+
+    It assumes that the ID column uniquely identifies the event initiator and
+    their counterparts' event. Choose only tables for which this assumption is
+    true. In some cases, asynchronous communication like SMS might not be
+    tagged with an ID that allows one to recover the counterpart event.
+
+    Distances are measured in km.
+
+    Parameters
+    ----------
+    start, stop : str
+         iso-format start and stop datetimes
+    hours : 2-tuple of floats, default 'all'
+        Restrict the analysis to only a certain set
+        of hours within each day.
+    tables: str, default 'all'.
+        The table must have a `msisdn_counterpart` column.
+    subscriber_identifier : {'msisdn', 'imei'}, default 'msisdn'
+        Either msisdn, or imei, the column that identifies the subscriber.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    statistic :  {'count', 'sum', 'avg', 'max', 'min', 'median', 'mode', 'stddev', 'variance'}, default 'avg'
+        Defaults to sum, aggregation statistic over the durations.
+
+
+    Examples
+    --------
+
+    >>> s = DistanceCounterparts("2016-01-01", "2016-01-07", statistic="avg")
+    >>> s.get_dataframe()
+
+              subscriber    distance_avg
+        038OVABN11Ak4W5P      272.167815
+        09NrjaNNvDanD8pk      241.290233
+        0ayZGYEQrqYlKw6g      218.161568
+        0DB8zw67E9mZAPK2      228.235324
+        0Gl95NRLjW2aw8pW      189.008980
+                     ...             ...
+
+    """
+
+    def __init__(
+        self,
+        start,
+        stop,
+        statistic="avg",
+        *,
+        hours="all",
+        tables="all",
+        direction="both",
+        exclude_self_calls=True,
+        subscriber_subset=None,
+    ):
+        self.tables = tables
+        self.start = start
+        self.stop = stop
+        self.hours = hours
+        self.direction = direction
+        self.exclude_self_calls = exclude_self_calls
+
+        self.statistic = statistic.lower()
+        if self.statistic not in valid_stats:
+            raise ValueError(
+                "{} is not a valid statistic. Use one of {}".format(
+                    self.statistic, valid_stats
+                )
+            )
+
+        column_list = ["msisdn", "msisdn_counterpart", "id", "location_id", "outgoing"]
+        self.tables = parse_tables_ensuring_columns(
+            self.connection, tables, column_list
+        )
+
+        # EventsTablesUnion will only subset on the subscriber identifier,
+        # which means that we need to query for a unioned table twice.
```

Comment:
I haven't had time to dig into this in detail, but my hunch is that the double join between the two unioned tables (in `_make_query` below) is likely to be very slow. The test currently takes ~12 seconds to run, and this is for the small test data! Have you had a chance to test this on larger data, or do you see a way to improve this? It's not a blocker if not, but we should leave a note somewhere to indicate the performance, and revisit it later.

Reply:
Haven't done this test, but I agree that it should slow things down given that it is filtering the `events` table twice. Do you think it would perhaps more sense to simple join to the unfiltered `event` table using `msisdn_counterpart` as the key?

Ideally, I would like to see the `subscriber_subset` also collecting events for which the `msisdn_counterpart` is present. At the moment, it only allows for one column to be used for filtering.

Reply:
Yeah, might make sense to test alternative methods to calculate the result, to see if any speed-ups can be achieved. I'd say it's not needed for this PR, but we should look into it.